torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
   throw std::runtime_error("ProcessGroupMPI does not support reduce_scatter");
 }
 
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::alltoall_base(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    std::vector<int>& outputSplitSizes,
+    std::vector<int>& inputSplitSizes,
+    const AllToAllOptions& opts) {
+  checkSingleTensor(outputTensors);
+  checkSingleTensor(inputTensors);
+
+  if(outputSplitSizes.size() == 0 && inputSplitSizes.size() == 0) {
+    // We can use alltoall
+    if ((outputTensors[0].numel() != inputTensors[0].numel()) ||
+        (outputTensors[0].type() != inputTensors[0].type())) {
+      throw std::runtime_error("Tensors are not equal in size or data type");
+    }
+    checkSingleTensorHelper(inputTensors[0]);
+    checkSingleTensorHelper(outputTensors[0]);
+    if(outputTensors[0].size(0) % size_ != 0) {
+      throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+    }
+    
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoall(
+              srcdata.data_ptr(),
+              srcdata.numel()/size_,
+              mpiDatatype.at(srcdata.scalar_type()),
```

Comment:
MPI and PyTorch types may not have a 1:1 correspondence (for example, PyTorch supports float16 which MPI does not). One way to fix this for functions that only move data is to set MPI type to MPI_BYTE and multiply the counts by `sizeof(PyTorchElementType)`. Alternatively, create a 1:N map that maps PyTorch types to the MPI type with the same in-memory size. For example, map float16 to MPI_SHORT. Just be careful not to pass these into computational functions like `MPI_Reduce` or `MPI_Scan` otherwise the results will be incorrect.  I prefer the former method since it's a bit more future-proof but it does have the downside that count in byte needs to fit into int32.

Reply:
@drdarshan The question of how float16 is handled by MPI for Reduce/Allreduce is a critical issue that should determine how we handle datatypes in general for functions that only move data. We will need to rely on other backends, e.g. NCCL or Intel CCL, since stock MPI is not expected to support this anytime soon. What is the guidance here? Is it ok for float16 to get mapped to work with alltoall/allgather but fail with allreduce when using MPI backend? Might be better to do some checks for these new data-types at init and see if the backend supports it or not.

Reply:
Here is my quick fix commit for adding support for Half and BFloat16 datatypes in MPI backend https://github.com/ddkalamk/pytorch/commit/34e21c9fc9fadb2167ef64036314898df52190a8

Reply:
@srinivas212 - just my two cents but pure data movement operations should not in general error out for different data types. On the other hand, some computational operations might just be undefined or infeasible to implement for some (type, backend) combination. Even with MPI, MAX or MAXLOC with Complex and Double Complex is not supported in the standard. In such cases, I'd just suggest documenting these cases and failing fast.  
@ddkalamk - Your fix looks good to me! My only suggestion would be to use the `MPI_CHECK` macro as is done with other calls. It's highly unlikely that calls like `MPI_Op_create` will fail but it might be good to be consistent with the rest of the code.

Reply:
> My only suggestion would be to use the MPI_CHECK macro as is done with other calls.

Agree with you. As I said in the comment, it's a quick fix commit to test something :). But will fix it.